File: android/app/src/main/java/com/rcforb/android/services/LobbyService.kt
package com.rcforb.android.services

import com.rcforb.android.models.RemoteStation
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import java.util.regex.Pattern

object LobbyService {
    private const val FEED_URL = "http://online.remotehams.com/xmlfeed.php"
    private val client = OkHttpClient()

    suspend fun fetchStations(): List<RemoteStation> = withContext(Dispatchers.IO) {
        try {
            val request = Request.Builder().url(FEED_URL).build()
            val response = client.newCall(request).execute()
            val xml = response.body?.string() ?: ""
            parseStationXML(xml)
        } catch (e: Exception) {
            android.util.Log.e("LobbyService", "Fetch error", e)
            emptyList()
        }
    }

    private fun parseStationXML(xml: String): List<RemoteStation> {
        val stations = mutableListOf<RemoteStation>()
        val pattern = Pattern.compile("<Radio>([\\s\\S]*?)</Radio>")
        val matcher = pattern.matcher(xml)

        while (matcher.find()) {
            val block = matcher.group(1) ?: continue
            parseRadioBlock(block)?.let { stations.add(it) }
        }
        return stations
    }

    private fun getField(block: String, tag: String): String {
        val pattern = Pattern.compile("<$tag>(.*?)</$tag>", Pattern.DOTALL)
        val matcher = pattern.matcher(block)
        return if (matcher.find()) matcher.group(1)?.trim() ?: "" else ""
    }

    private fun parseRadioBlock(block: String): RemoteStation? {
        val orbId = getField(block, "OrbId")
        val domain = getField(block, "Domain")
        if (orbId.isEmpty() || domain.isEmpty()) return null

        val port = getField(block, "Port").toIntOrNull() ?: 4525
        val voipPort = getField(block, "VoipPort").toIntOrNull() ?: 4524
        val serverName = getField(block, "ServerName").ifEmpty { "Unknown" }

        return RemoteStation(
            serverId = orbId,
            serverName = serverName,
            description = getField(block, "Message"),
            host = domain,
            port = port,
            voipPort = voipPort,
            online = getField(block, "Online").lowercase() == "true",
            radioInUse = false,
            radioOpen = true,
            serverVersion = getField(block, "Version"),
            radioModel = getField(block, "RadioName"),
            country = getField(block, "Country"),
            gridSquare = getField(block, "Grid"),
            latitude = getField(block, "Latitude").toDoubleOrNull() ?: 0.0,
            longitude = getField(block, "Longitude").toDoubleOrNull() ?: 0.0,
            userCount = getField(block, "Users").toIntOrNull() ?: 0,
            maxUsers = getField(block, "MaxUsers").toIntOrNull() ?: 0,
            isV7 = false
        )
    }
}

File: ipadOS/RCForb/Sources/Services/LobbyService.swift
import Foundation

enum LobbyService {
    private static let feedURL = "http://online.remotehams.com/xmlfeed.php"

    static func fetchStations() async -> [RemoteStation] {
        guard let url = URL(string: feedURL) else { return [] }

        do {
            let (data, _) = try await URLSession.shared.data(from: url)
            let xml = String(data: data, encoding: .utf8) ?? ""
            return parseStationXML(xml)
        } catch {
            print("[LobbyService] Fetch error: \(error)")
            return []
        }
    }

    private static func parseStationXML(_ xml: String) -> [RemoteStation] {
        var stations: [RemoteStation] = []

        let pattern = "<Radio>([\\s\\S]*?)</Radio>"
        guard let regex = try? NSRegularExpression(pattern: pattern) else { return [] }
        let matches = regex.matches(in: xml, range: NSRange(xml.startIndex..., in: xml))

        for match in matches {
            if let range = Range(match.range(at: 1), in: xml) {
                let block = String(xml[range])
                if let station = parseRadioBlock(block) {
                    stations.append(station)
                }
            }
        }
        return stations
    }

    private static func getField(_ block: String, _ tag: String) -> String {
        let pattern = "<\(tag)>(.*?)</\(tag)>"
        guard let regex = try? NSRegularExpression(pattern: pattern, options: .dotMatchesLineSeparators),
              let match = regex.firstMatch(in: block, range: NSRange(block.startIndex..., in: block)),
              let range = Range(match.range(at: 1), in: block) else { return "" }
        return String(block[range]).trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private static func parseRadioBlock(_ block: String) -> RemoteStation? {
        let orbId = getField(block, "OrbId")
        let domain = getField(block, "Domain")
        guard !orbId.isEmpty, !domain.isEmpty else { return nil }

        let port = Int(getField(block, "Port")) ?? 4525
        let voipPort = Int(getField(block, "VoipPort")) ?? 4524

        return RemoteStation(
            serverId: orbId,
            serverName: getField(block, "ServerName").isEmpty ? "Unknown" : getField(block, "ServerName"),
            description: getField(block, "Message"),
            host: domain,
            port: port,
            voipPort: voipPort,
            online: getField(block, "Online").lowercased() == "true",
            radioInUse: false,
            radioOpen: true,
            serverVersion: getField(block, "Version"),
            radioModel: getField(block, "RadioName"),
            country: getField(block, "Country"),
            gridSquare: getField(block, "Grid"),
            latitude: Double(getField(block, "Latitude")) ?? 0,
            longitude: Double(getField(block, "Longitude")) ?? 0,
            userCount: Int(getField(block, "Users")) ?? 0,
            maxUsers: Int(getField(block, "MaxUsers")) ?? 0,
            isV7: false
        )
    }
}
